File: ios/RibbonModule.m
#import <Foundation/Foundation.h>
#import "React/RCTBridgeModule.h"
@interface
RCT_EXTERN_MODULE(RibbonModule, NSObject)
RCT_EXTERN_METHOD(configure: (NSString*) name)
RCT_EXTERN_METHOD(triggerAndShow: (NSString*) value)
RCT_EXTERN_METHOD(trigger: (NSString*) name
    resolver:(RCTPromiseResolveBlock)resolve
    rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(showDialog)
RCT_EXTERN_METHOD(setPreview: (BOOL) value)
RCT_EXTERN_METHOD(getAttributes:
                  (RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(setAttributes: (NSDictionary*) attributes)
RCT_EXTERN_METHOD(removeAttribute: (NSString*) key)
RCT_EXTERN_METHOD(removeAttributes: (NSArray*) key)
RCT_EXTERN_METHOD(removeAllAttributes)
RCT_EXTERN_METHOD(setIdentity: (NSString*) userId)
RCT_EXTERN_METHOD(clearSession)
@end

File: ios/RibbonModule.swift
import Foundation
import AVFoundation
import RibbonKit
import OSLog


@objc(RibbonModule)
class RibbonModule: NSObject {
  

  @objc
  func configure(_ ribbonId: String) {
    Ribbon.shared.configure(withID: ribbonId)
  }
  
  @objc(trigger:resolver:rejecter:)
  func trigger(_ someValue: String,
               resolver resolve: @escaping RCTPromiseResolveBlock,
               rejecter reject: @escaping RCTPromiseRejectBlock) -> Void {
    Ribbon.shared.trigger(someValue) { state in
      print(state)
      switch state {
      case .noStudy:
        resolve("NoReady")
      case .ready:
        resolve("StudyReady")
      case .error(let error):
        reject("Error", "Failure", error)
      @unknown default:
        print("some other case")
      }
    }
  }
  
  @objc
  func triggerAndShow(trigger: String) {
    Ribbon.shared.triggerAndShow(trigger)
  }
  
  @objc
  func showDialog() {
    Ribbon.shared.show()
  }
  
  @objc(getAttributes:rejecter:)
  func getAttributes(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    let attr = Ribbon.shared.getAttributes()
    if (attr == nil){
      reject("Reject", "secondParam", nil)
    } else {
      resolve(attr)
    }
  }
  
  @objc
  func setAttributes(_ attributes: [String: Any]) {
    Ribbon.shared.setAttributes(attributes)
  }
  
  @objc
  func removeAttributes(key: [String]) {
    Ribbon.shared.removeAttributes(key)
  }
  
  @objc
  func removeAllAttributes() {
    Ribbon.shared.removeAllAttributes()
  }
  
  @objc
  func setIdentity(userId: String) {
    Ribbon.shared.setIdentity(userId)
  }
  
  @objc
  func setPreview(_ preview: Bool) {
    Ribbon.shared.setPreview(preview)
  }
  
  @objc
  func clearSession() {
    Ribbon.shared.clearSession()
  }
  
  @objc
  static func requiresMainQueueSetup() -> Bool {
    return true
  }
}
